{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "R6kBCqN6TEoj"
   },
   "source": [
    "---\n",
    "format: \n",
    "    html:\n",
    "      toc: true\n",
    "title: \"Bulk RNA Analysis\"\n",
    "author: \"Lavinia I Fechete, Stig U Andersen, Mikkel H Schierup, Samuele Soraggi\"\n",
    "title-block-banner: true\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "0jp4jvtHS2LW"
   },
   "source": [
    ":::{.callout-note title=\"Tutorial description\"}\n",
    "\n",
    "This tutorial will cover the steps for performing Differential Gene Expression on the RNA-seq data obtained from Galaxy. At the end of this tutorial you will be able to use `R` to\n",
    "    \n",
    "- Perform preliminary analyses of the RNA-seq results\n",
    "- Find differentially expressed genes between two conditions using `edgeR`\n",
    "\n",
    ":::\n",
    "\n",
    "The present tutorial, like the rest of the course material, is available at our [open-source github repository](https://github.com/hds-sandbox/NGS_summer_course_Aarhus). \n",
    "\n",
    "To use this notebook, use the `NGS (R)` kernel that contains the packages. Choose it by selecting `Kernel -> Change Kernel` in the menu on top of the window.\n",
    "\n",
    ":::{.callout-tip collapse=\"true\" title=\"A few introductory points to run this notebook (click to show)\"}\n",
    "  \n",
    "* To use this notebook, use the `NGS (R)` kernel that contains the packages. Choose it by selecting `Kernel -> Change Kernel` in the menu on top of the window.\n",
    "\n",
    "<figure>\n",
    "<img src=\"images/kernelchoice.png\" width=\"500\" alt=\"Kernel Choice\" class=\"center\">\n",
    "</figure>\n",
    "\n",
    "* In this notebook you will use only R commands\n",
    "* On some computers, you might see the result of the commands once they are done running. This means you will wait some time while the computer is crunching, and only afterwards you will see the result of the command you have executed\n",
    "*  You can run the code in each cell by clicking on the run cell button, or by pressing <kbd> Shift </kbd> + <kbd> Enter </kbd>.  When the code is done running, a small green check sign will appear on the left side\n",
    "*   You need to run the cells in sequential order, please do not run a cell until the one above finished running and do not skip any cells\n",
    "*   Each cell contains a short description of the code and the output you should get. Please try not to focus on understanding the code for each command in too much detail, but rather try to focus on the output \n",
    "*   You can create new code cells by pressing <kbd> + </kbd> in the Menu bar above. \n",
    "    \n",
    ":::\n",
    "   \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ByrhPfLnV6v5"
   },
   "source": [
    "**Load the necessary R libraries**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "executionInfo": {
     "elapsed": 9,
     "status": "ok",
     "timestamp": 1654592586765,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "wqF_8OxWtx3M",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(VennDiagram)\n",
    "library(dplyr)\n",
    "library(tibble)\n",
    "library(formattable)\n",
    "library(mixOmics)\n",
    "library(pheatmap)\n",
    "library(edgeR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VfiGz1s3mljb",
    "tags": []
   },
   "source": [
    "# EdgeR - Data filtering and Normalization \n",
    "\n",
    "We will use the `edgeR` package to test for differentially expressed genes between our Control and Treatment Samples. You can read more about the edgeR package and its functionalities [here](https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf). This exercise largely follows their user manual. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-07T14:12:04.188869Z",
     "iopub.status.busy": "2022-06-07T14:12:04.185495Z",
     "iopub.status.idle": "2022-06-07T14:12:04.203702Z",
     "shell.execute_reply": "2022-06-07T14:12:04.201977Z"
    }
   },
   "source": [
    "## File processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "GI0DcWp7uoE6"
   },
   "source": [
    "The data for this exercise comes from the 12 tabular files with Reads per Gene counts generated by STAR Mapping in the `raw-data alignment` part of this course.\n",
    "We want to create a table where each column is a sample, and the content of the table are the read counts from STAR. We must merge the 12 files with Reads per Gene information into a single file.\n",
    "\n",
    "* If you aligned datasets in the first notebook with jupyterlab, then you will find the files using the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "samples <- sort(system(\"find results/STAR_output/*_align_contigs_1_2 -name \\\"*ReadsPerGene.out.tab\\\"\", intern=TRUE))\n",
    "print(samples)\n",
    "Read_counts <- do.call(cbind, lapply(samples, function(x) read.delim(file=x, header = FALSE)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* If you aligned datasets interactively with `Galaxy`, then you will need to\n",
    "    - create the folder `tabular_files` into the `results` folder\n",
    "    - copy the `tabular` files from `STAR` into the created folder. Each file must have the sample name and end by `ReadsPerGene.out.tab`. For example `S10_1_1ReadsPerGene.out.tab`\n",
    "    - run the following commands removing the `#` symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#samples <- sort(system(\"find results/tabular_files/ -name \\\"*ReadsPerGene.out.tab\\\"\", intern=TRUE))\n",
    "#print(samples)\n",
    "#Read_counts <- do.call(cbind, lapply(samples, function(x) read.delim(file=x, header = FALSE)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data frame has genes as rows and samples as columns and stores the gene expression counts (value representing the total number of sequence reads that originated from a particular gene in a sample) for each of the 12 samples. This data frame should have 12 columns and 366 rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 728,
     "status": "ok",
     "timestamp": 1654592226501,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "EnaYAZR6t6pG",
    "outputId": "695957c4-1210-4c5c-c576-5addfe921892",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "rownames(Read_counts) <- Read_counts[,1]\n",
    "Read_counts <- Read_counts[c(5:nrow(Read_counts)), c(seq(2, 46, by=4))]\n",
    "colnames(Read_counts) <- c(\"S10_1_1\", \"S10_1_2\", \"S10_1_3\", \"S10_2_1\", \"S10_2_2\", \"S10_2_3\", \n",
    "                       \"TI_1_1\", \"TI_1_2\", \"TI_1_3\", \"TI_2_1\", \"TI_2_2\", \"TI_2_3\")\n",
    "head(Read_counts, n=10)\n",
    "dim(Read_counts) # dimensions of the data frame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "i_abVRLburz7"
   },
   "source": [
    "Import the Metadata table. This file, as its name suggests, contains information about each of the 12 RNA-seq samples, such as the treatment (Condition), genotype and replicate.\n",
    "\n",
    "Note that the order of the rows in the Metadata table should be the same as the columns in the Read_counts file generated above. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 799,
     "status": "ok",
     "timestamp": 1654592230356,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "Lkl15eGwn9aq",
    "outputId": "98c192c7-46f3-4662-819b-81cddaafb3b7",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "metadata <- read.csv(\"../Data/Clover_Data/metadata.csv\", sep =\";\", row.names=1, stringsAsFactors=TRUE)\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "GGSMFInHhBur"
   },
   "source": [
    "In order to aid the following steps, we will create a Group for each sample (a new column in the metadata) based on the Genotype&Condition of each sample and assign the three replicates to this group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 440,
     "status": "ok",
     "timestamp": 1654592249297,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "TTrXR2d8FgTK",
    "outputId": "7b78e7c3-a246-4638-9e56-68605e85ad52",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "Group <- factor(paste(metadata$Genotype, metadata$Condition, sep=\"_\"))\n",
    "metadata <- cbind(metadata,Group=Group)\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "IBNXxE0PLHCz"
   },
   "source": [
    "## Create the DGEList object\n",
    "We will merge the read counts and the metadata into a list-based data object named DGEList, which can be manipulated as any list object in R.\n",
    "\n",
    "The main components of the DGEList object are the matrix \"counts\" containing our read per gene counts and a data.frame \"samples\" containing the metadata. \n",
    "\n",
    "Note that all the genes with zero counts across all samples were eliminated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 896,
     "status": "ok",
     "timestamp": 1654592267233,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "97YnGnzeoFAu",
    "outputId": "64d6fedd-0394-4edc-ca79-519e5d30de63",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "DGEList <- DGEList(Read_counts, remove.zeros = TRUE)\n",
    "DGEList$samples$Condition <- relevel(metadata$Condition, ref = \"Control\")\n",
    "DGEList$samples$Genotype <- metadata$Genotype\n",
    "DGEList$samples$group <- metadata$Group\n",
    "DGEList"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "5ZCIbthBuzBI"
   },
   "source": [
    "## Preliminary data analysis\n",
    "First, we will calculate the \"pseudoCounts\" as log2 of the reads per gene counts. \n",
    "<br>\n",
    "This is not part of the actual differential gene expression analysis but is helpful for data exploration and quality assessment. We will look at a histogram of one of the samples and a boxplot representation of the log2 counts for all the 12 samples.\n",
    "<br>\n",
    "Note that there are many genes with a low number of mapped reads and that there are differences between the average read counts for each library. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "executionInfo": {
     "elapsed": 853,
     "status": "ok",
     "timestamp": 1654592276616,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "uEiv2JpKqDiz",
    "outputId": "b0938de1-e0c6-44cc-a721-2f6fb5298908",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "pseudoCounts <- log2(DGEList$counts + 1)\n",
    "head(pseudoCounts)\n",
    "hist(pseudoCounts[ ,\"S10_1_1\"], main = \"\", xlab = \"Read counts\")\n",
    "boxplot(pseudoCounts, col = \"gray\", las = 3, cex.names = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "KQjh9mrC86_M"
   },
   "source": [
    "We can also create a PCA plot of the samples in order to assess the differences between the Genotypes and Conditions, but also between the replicates. In this plot the samples that are similar cluster together, while samples that are different are further apart.\n",
    "<br>\n",
    "In this type of plot, we would expect samples from the same group (the three replicates for each sample) to exhibit a similar gene expression profile thus clustering together while being separated from the other samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 497
    },
    "executionInfo": {
     "elapsed": 892,
     "status": "ok",
     "timestamp": 1654592287896,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "b_zFi1eSSi8v",
    "outputId": "9e9c75ec-eaff-4de8-df49-e710d7723f6f",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "resPCA <- pca(t(pseudoCounts), ncomp = 6)\n",
    "plotIndiv(resPCA, group = metadata$Genotype, pch=metadata$Condition,\n",
    "                  legend = T, legend.title = 'Genotype', legend.title.pch = 'Condition',\n",
    "                  title = 'PCA plot raw counts', style = 'ggplot2', size.xlabel = 10, size.ylabel = 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "hrEmdocWHieJ"
   },
   "source": [
    "## Filtering the lowly expressed genes\n",
    "As seen previously, many genes have a low number of read counts in our samples. The genes with very low counts across all libraries provide little evidence for differential expression, thus we should eliminate these genes before the analysis. \n",
    "<br>\n",
    "One of the filtering methods we can use is the \"filterByExpr\" function provided by the edgeR package. By default, this function keeps only the genes that have at least 10 reads per group, but other cutoffs can also be applied."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 302,
     "status": "ok",
     "timestamp": 1654592373252,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "cefMmTaPuSLn",
    "outputId": "6aaf80e9-b0b5-493d-d217-a6233ffda53f",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "keep <- filterByExpr(DGEList, group=metadata$Group) #create the filter\n",
    "DGEList <- DGEList[keep, , keep.lib.sizes=FALSE] #apply the filter to on the DGEList object\n",
    "table(keep) #Check the number of genes that passed the filter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "fRhZxuchv8d_"
   },
   "source": [
    "## Normalization\n",
    "As we are working with multiple samples we need to normalize the read counts per gene in order to account for compositional and technical differences between the 12 RNA-seq libraries. For this, we will calculate normalization factors using the trimmed mean of M-values (TMM) method. You can read more about different normalization methods in the user manual.  \n",
    "<br>\n",
    "Note that running \"calcNormFactors\" does not change the actual reads per gene counts, it just fills the \"norm.factors\" column in `DGEList$samples`. \n",
    "\n",
    "These factors will be used to scale the read counts for each library.\n",
    "From the user guide: \"A normalization factor below one indicates that a small number of high count genes\n",
    "are monopolizing the sequencing, causing the counts for other genes to be lower than would\n",
    "be usual given the library size.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 706,
     "status": "ok",
     "timestamp": 1654592376753,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "R9W-K1DYraK_",
    "outputId": "380c972e-581e-46e0-aba4-f79032597c21",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "DGEList <- calcNormFactors(DGEList, method=\"RLE\")\n",
    "DGEList$samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "AgOkkvTQrxaX",
    "tags": []
   },
   "source": [
    "## Normalized counts - Exploratory Data analysis\n",
    "For data analysis purposes normalized log2 counts can be extracted from the DGEList object using the function CPM (counts per million).\n",
    "\n",
    "We will generate the same plots as for the raw counts in order to compare the data before and after normalization. \n",
    "\n",
    "Do the plots for normalized counts look different compared with the plots computed before data filtering and normalization? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "executionInfo": {
     "elapsed": 912,
     "status": "ok",
     "timestamp": 1654592379673,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "-JCT9WcqrwIu",
    "outputId": "7def7808-5abd-466f-ed36-5a4706348a25",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "pseudoNormCounts <- cpm(DGEList, log = TRUE, prior.count = 1)\n",
    "head(pseudoNormCounts)\n",
    "hist(pseudoNormCounts[ ,\"S10_1_1\"], main = \"\", xlab = \"counts\")\n",
    "boxplot(pseudoNormCounts, col = \"gray\", las = 3, cex.names = 1)\n",
    "\n",
    "resPCA <- pca(t(pseudoNormCounts), ncomp = 6)\n",
    "plotIndiv(resPCA, group = metadata$Genotype, pch=metadata$Condition,\n",
    "                  legend = T, legend.title = 'Genotype', legend.title.pch = 'Condition',\n",
    "                  title = 'PCA plot normalized counts', style = 'ggplot2', size.xlabel = 10, size.ylabel = 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MHgVaHK2xr0N"
   },
   "source": [
    "# EdgeR - Testing for Differentially expressed genes (DEGs)\n",
    "\n",
    "We use `edgeR` for Differential Gene Expression using the GLM method (Generalized linear model) with treatment and genotype effects\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "T2jGcTIoAaf4"
   },
   "source": [
    "After we concluded the exploratory data analysis and the filtering and normalization steps we can now start testing for differentially expressed genes between our samples. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "HlvrtfEX4CCS"
   },
   "source": [
    "For this, we first need to define a \"design matrix\" which describes our experimental design. We construct a matrix containing samples as rows and our previously defined groups (S10_Control, S10_Treatment, Tienshan_Control, Tienshan_Treatment) as columns. The samples belonging to each group are assigned a \"1\" in this matrix.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 738,
     "status": "ok",
     "timestamp": 1654592391013,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "HYD9kUpSrxMG",
    "outputId": "214fd833-17b3-465c-de56-6f60416ab515",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "design.matrix <- model.matrix(~0+Group)\n",
    "rownames(design.matrix) <- colnames(DGEList)\n",
    "colnames(design.matrix) <- levels(metadata$Group)\n",
    "design.matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Td0C_CdS4lov"
   },
   "source": [
    "For easier pairwise comparisons between the groups we can create some \"contrasts\" using the \"makeContrasts\" function. This will tell the program which samples we will like to consider as our \"Control\" and which as our \"Treatment\" groups. The Control groups will be assigned a \"-1\" and the Treatment groups a \"1\". When using multiple Control and Treatment samples in the same contrast we need to devide by the number of samples used as for example in the contrast S10_Tienshan, where we use both S10 and Ti samples.\n",
    "\n",
    "For example, the contrast `S10 = S10_Treatment - S10_Control` will test for differentially expressed genes between the S10 treatment samples vs the S10 Control samples. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 291,
     "status": "ok",
     "timestamp": 1654592481861,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "QyfOq6Yc75gA",
    "outputId": "e7164a60-f51b-4349-85d7-4209efd7b617",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "contrasts <- makeContrasts(\n",
    "                S10 = S10_Treatment-S10_Control,\n",
    "                Tienshan = Tienshan_Treatment-Tienshan_Control,\n",
    "                S10_Tienshan= (S10_Treatment+Tienshan_Treatment)/2 - (S10_Control+Tienshan_Control)/2,\n",
    "                levels=design.matrix)\n",
    "contrasts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "rTfNfL_DVPL4"
   },
   "source": [
    "*   *Any ideas for other contrasts that migth be interesting to explore?*\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "B3bw7RR6h7YB"
   },
   "source": [
    "The following cell will run two mandatory steps in order to fit a model in edgeR. First, the dispersion needs to be estimated, this is a measure of the degree of inter-library variation for each gene. Afterwards, a QL ( quasi-likelihood) model representing the sample design is fitted using the \"glmQLFit\" function. You can read more about these steps in the edgeR user guide."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "executionInfo": {
     "elapsed": 592,
     "status": "ok",
     "timestamp": 1654592493067,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "zM841xCPyFIJ",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "DGEList <- estimateDisp(DGEList, design.matrix)\n",
    "fit <- glmQLFit(DGEList, design.matrix)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "7tzzIZjq7Qcb"
   },
   "source": [
    "## DEGs for White clover S10 plants Treatment vs Control\n",
    "\n",
    "Find DEGs for White clover S10 plants in Treatment condition compared with the Control condition\n",
    "We will test for differentially expressed genes using the quasi-likelihood F-tests method. This is easily done just by running the \"glmQFTest\" function on the fit model and selecting one of the contrasts we created above (in this case the \"S10\" contrast).\n",
    "\n",
    "Next, we will use the topTags function to extract the information about all the genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 3844,
     "status": "ok",
     "timestamp": 1654592502170,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "6vOoqgc_V3EI",
    "outputId": "98005ba9-3ece-4af3-dbbb-bda6f571f6fe",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "glmqlf_S10 <- glmQLFTest(fit, contrast=contrasts[,\"S10\"])\n",
    "DEG_S10 <- topTags(glmqlf_S10, n = nrow(DGEList$counts))\n",
    "DEG_S10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "GQ0RYFIXzb33"
   },
   "source": [
    "We have now created a table with certain parameters calculated for each of the genes analysed.\n",
    "\n",
    "**logFC** represents the base 2 logarithm of the fold change and it shows us how much the expression of the gene has changed between the two conditions. A logFC of 1 means a doubling in the read per gene count between the control and treatment samples.\n",
    "The genes with a logFC higher than 0 are upregulated while the genes with a logFC lower than 0 are downregulated between the control and treatment samples. \n",
    "<br>\n",
    "**logCPM** represents the average log2-counts-per-million, the abundance of the gene.\n",
    "\n",
    "**F** - F-statistic.\n",
    "\n",
    "**PValue** is the raw p-value.\n",
    "\n",
    "**FDR** (The false discovery rate) represents the adjusted p-value and is calculated using Benjamini and Hochberg's algorithm. It controls the rate of false positive values under multiple testing. Usually, a threshold of under 5% is set for the FDR.  \n",
    "\n",
    "The important information for us in this table is stored in the \"logFC\" and the \"FDR\" columns. The top DE genes have small FDR values and large fold changes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "qaneQKfFaYir"
   },
   "source": [
    "Many of the genes in the samples are uninteresting for us, as they have a high FDR and/or low logFC values so we cannot consider them as differentially expressed.\n",
    "\n",
    "We will apply a filtering step in order to keep only the statistically significant genes. We will filter out the genes with an FDR higher than 0.05 and an absolute logFC lower than 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 678,
     "status": "ok",
     "timestamp": 1654592518729,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "U7s5TKLuuhlo",
    "outputId": "7cebe4ec-b387-44a6-a992-6544f44d8803",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "DEG_S10_filtered <- DEG_S10$table[DEG_S10$table$FDR < 0.05 & abs(DEG_S10$table$logFC) > 1,] #Filtering\n",
    "DEG_S10_filtered <- rownames_to_column(DEG_S10_filtered) %>% rename(gene_ID = rowname) #Adding the gene_ID column\n",
    "head(DEG_S10_filtered)\n",
    "nrow(DEG_S10_filtered) # finding the number of genes that passed the filter "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "apC30py88Tln"
   },
   "source": [
    "We can also visualize the selected genes by plotting a Smear plot or a Volcano plot. The Genes that passed the filter are coloured in red, and the top 10 genes with the lowest FDR value are labelled with their gene ID.\n",
    "\n",
    "We can see that the majority of the genes analysed are either not statistically significant or have a very small logFC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 497
    },
    "executionInfo": {
     "elapsed": 335,
     "status": "ok",
     "timestamp": 1654592521575,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "RrjE2BhW0z-e",
    "outputId": "15318014-668e-4b38-dd94-ddb9bb81f396",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "plotSmear(glmqlf_S10,\n",
    "          de.tags = rownames(DEG_S10$table)[which(DEG_S10$table$FDR < 0.05 & abs(DEG_S10$table$logFC) > 1)])\n",
    "text(x=DEG_S10_filtered$logCPM[1:10],\n",
    "     y=DEG_S10_filtered$logFC[1:10],\n",
    "     labels=DEG_S10_filtered$gene_ID[1:10], cex=0.7, pos=1)\n",
    "abline(h = c(-1, 1), col = \"blue\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "k9J6gJ9iUtOf"
   },
   "source": [
    "We can also create a heatmap with the log2 read counts of the selected differentially expressed genes so that we can visualise the differences in normalized counts between the Control and the Treatment samples.\n",
    "\n",
    "The genes are ordered by the FDR value. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 497
    },
    "executionInfo": {
     "elapsed": 903,
     "status": "ok",
     "timestamp": 1654592598499,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "IxhY-fTVFt8i",
    "outputId": "340a6abb-dfc0-433f-f987-8d79f7ed0437",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "annot_col <- data.frame(row.names = colnames(pseudoNormCounts)[1:6], Condition = c(rep(\"Control\", 3),rep( \"Treatment\", 3)))                   \n",
    "pheatmap(as.matrix(pseudoNormCounts[DEG_S10_filtered$gene_ID,c(1:6)]), cluster_rows = F, cluster_col = F, annotation_col = annot_col)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Dpo0xoqzKFVA"
   },
   "source": [
    "## DEGs for White clover Tienshan Treatment vs Control\n",
    "We can do this using the same functions as above and changing the contrast.\n",
    "\n",
    "This time we will directly filter the differentially expressed genes using the same parameters as for the S10 samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 299,
     "status": "ok",
     "timestamp": 1654592628511,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "XHfwYQw2Z_fB",
    "outputId": "22bd389e-fc28-43c6-ff33-e464164c06a6",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "glmqlf_Ti <- glmQLFTest(fit, contrast=contrasts[,\"Tienshan\"])\n",
    "DEG_Ti <- topTags(glmqlf_Ti, n = nrow(DGEList$counts))\n",
    "DEG_Ti_filtered <- DEG_Ti$table[DEG_Ti$table$FDR < 0.05 & abs(DEG_Ti$table$logFC) > 1,]\n",
    "DEG_Ti_filtered <- rownames_to_column(DEG_Ti_filtered) %>% rename(gene_ID = rowname)\n",
    "print(head(DEG_Ti_filtered))\n",
    "print(\"Nr of differentially expressed genes:\")\n",
    "print(nrow(DEG_Ti_filtered))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "x3k2KcekDTyA"
   },
   "source": [
    "We can now plot a Venn diagram with the DEGs for the two genotypes in order to observe the number of common and specific differentially expressed genes between the two genotypes as response to the cold exposure. You can see that a high percentage of the identified genes are common for the two genotypes, while each genotype has also specific genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 497
    },
    "executionInfo": {
     "elapsed": 455,
     "status": "ok",
     "timestamp": 1654592637394,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "7Gzzzf8JTvWZ",
    "outputId": "72c86e42-33b5-4a6e-e3b3-30816eb231b6",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "vd <- venn.diagram(\n",
    "  x = list(DEG_S10_filtered$gene_ID, DEG_Ti_filtered$gene_ID),\n",
    "  category.names = c(\"S10\" , \"Tienshan\"),\n",
    "  lwd = 4,\n",
    "  fill = c(\"cornflowerblue\", \"yellowgreen\"),\n",
    "  filename = NULL,\n",
    "  cat.cex = 1,\n",
    "  cat.fontface = \"bold\",\n",
    "  output=TRUE\n",
    ")\n",
    "grid.draw(vd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "rk-Jav3xKUjG"
   },
   "source": [
    "## DEGs for S10+Tienshan Treatment vs Control w/o genotype effects\n",
    "\n",
    "Until now we tested for DEGs specific for each of the two genotypes under cold treatment. We can also run a test where we ignore the genotype and just test for the differences in the cold response.\n",
    "<br>\n",
    "Consider the counts for both genotypes as a single dataset using the previously created contrast \"S10_Tienshan\". \n",
    "<br>\n",
    "* Do the results look different compared with the previous tests?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 705,
     "status": "ok",
     "timestamp": 1654592959246,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "bUNUnvONaGfD",
    "outputId": "9c18f993-436d-4f83-e440-339e6b267631",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "glmqlf_S10_Ti <- glmQLFTest(fit, contrast=contrasts[,\"S10_Tienshan\"])\n",
    "DEG_S10_Ti <- topTags(glmqlf_S10_Ti, n = nrow(DGEList$counts))\n",
    "DEG_S10_Ti_filtered <- DEG_S10_Ti$table[DEG_S10_Ti$table$FDR < 0.05 & abs(DEG_S10_Ti$table$logFC) > 1,]\n",
    "DEG_S10_Ti_filtered <- rownames_to_column(DEG_S10_Ti_filtered)  %>% rename(gene_ID = rowname)\n",
    "print(head(DEG_S10_Ti_filtered))\n",
    "print(\"Nr of differentially expressed genes:\")\n",
    "print(nrow(DEG_S10_Ti_filtered))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "5wOTlT7HDyJj"
   },
   "source": [
    "Plot the results from the 3 tests in a Venn diagram to visualize the number of common and unique genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 497
    },
    "executionInfo": {
     "elapsed": 464,
     "status": "ok",
     "timestamp": 1654592962509,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "Zt7hF5o_XEPU",
    "outputId": "df9b7407-c040-4c1d-b480-312522f8b1a7",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "vd <- venn.diagram(\n",
    "  x = list(DEG_S10_filtered$gene_ID, DEG_Ti_filtered$gene_ID,\n",
    "           DEG_S10_Ti_filtered$gene_ID),\n",
    "  category.names = c(\"S10\" , \"Tienshan\", \"S10_Tienshan\"),\n",
    "  lwd = 3,\n",
    "  fill = c(\"cornflowerblue\", \"yellowgreen\", \"thistle3\"),\n",
    "  filename = NULL,\n",
    "  cat.cex = 1,\n",
    "  cat.fontface = \"bold\",\n",
    "  output=TRUE\n",
    ")\n",
    "grid.draw(vd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Explore DGE results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "zdFlsLILa7e3"
   },
   "source": [
    "### Select the genes which appear only in the analysis using both genotypes for further examination.\n",
    "\n",
    "We can use the \"anti_join\" function from the dplyr package to keep only the unique genes that appear only when using both genotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "9EWR08KmHYMB",
    "outputId": "f1cb0416-2fc8-4d25-c378-3712e7cac131",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "S10_Ti_unique <- anti_join(DEG_S10_Ti_filtered, DEG_S10_filtered, by=\"gene_ID\") %>%\n",
    "                       anti_join(DEG_Ti_filtered, by=\"gene_ID\")\n",
    "S10_Ti_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "06oKpDFZInrD"
   },
   "source": [
    "### Linking the genes selected as differentially expressed back to the raw read counts\n",
    "\n",
    "How do the counts look for these genes, does it make sense that they are differentially expressed only when using the two genotypes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "4SKJ727zOKRB",
    "outputId": "495b2bbd-e625-4c4c-95c0-5e6fba3ed677",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "Read_counts <- rownames_to_column(Read_counts) %>% rename(gene_ID = rowname)\n",
    "S10_Ti_unique_counts <- inner_join(S10_Ti_unique[,c(1,2,6)], Read_counts, by=\"gene_ID\")\n",
    "S10_Ti_unique_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MmwdKeQyN-hg"
   },
   "source": [
    "### Adding functional annotation to DEGs\n",
    "Until now we looked only at gene IDs, but we can also add functional annotations to the DEGs. The functional annotations were generated using protein sequences and the `EggNOG` software.\n",
    "\n",
    "Identifying the molecular function of the differentially expressed genes can help us do a literature survey in order to check if any of the genes discovered have been previously identified as being involved in the cold response."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "cecgm4MVWAlp",
    "outputId": "30170679-e42b-4184-c2fd-5451dfc656ae",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "Functional_annotations <- read.delim(\"../Data/Clover_Data/Functional_Annotations.txt\")\n",
    "S10_Ti_unique_FA <- inner_join(S10_Ti_unique[,c(1,2,6)], Functional_annotations, by=\"gene_ID\")\n",
    "S10_Ti_unique_FA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "06sty1R5YQsJ"
   },
   "source": [
    "<div class=\"alert-success\"> <font size=\"+2\"> <b> Tasks and Questions </b> </font> </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "F1f-dq2yYcCn"
   },
   "source": [
    "\n",
    "*   Based on the results obtained in the analysis so far, would you change the cut-off for the FDR and logFC to be more strict or more permissive? Look back at the raw counts for different FDR and logFC values and set the thresholds as you find appropriate.\n",
    "\n",
    "You can also plot histograms with the FDR and logFC values. \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 497
    },
    "id": "nRVBHbNzVczV",
    "outputId": "3ba78a4d-9651-40e1-b69b-55ee83104b77",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    " hist(DEG_S10$table$FDR , main = \"\", xlab = \"FDR\",  breaks= 200, xlim = range(c(0, 0.1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    " hist(DEG_S10$table$logFC , main = \"\", xlab = \"logFC\",  breaks= 50, xlim = range(c(-6, 6)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "TFOvBqoQaTFF"
   },
   "source": [
    "* Separate the upregulated and downregulated genes for each genotype and append functional annotations to them. \n",
    "* Identify the genes that are commonly upregulated in S10 and Tienshan samples and the uniquely upregulated genes for each genotype.\n",
    "* Why do you think some of the proteins appear in duplicates? <br>\n",
    "\n",
    "To answer these questions, it may be convenient to save summary tables from R and open them in excel. See the code below for examples of how to do this. Files can be downloaded by right-clicking on the file name.\n",
    "\n",
    "If you are familiar with R functions, you are welcome to use those for counting.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dir.create(\"DEG_Output_tables\", showWarnings = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 711
    },
    "executionInfo": {
     "elapsed": 461,
     "status": "error",
     "timestamp": 1654593190300,
     "user": {
      "displayName": "Lavinia Fechete",
      "userId": "03683080993527566677"
     },
     "user_tz": -120
    },
    "id": "bqyevMcc9CVo",
    "outputId": "4fca6a22-48cd-44d5-f221-f9172ea5ad3d",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Create the table with the DEGs, Raw counts(just for the Ti samples in this case, change to columns (2:7) for the S10 samples) and Functional annotations\n",
    "DEG_Ti_counts_FA <- inner_join(DEG_Ti_filtered[,c(1, 2, 6)], Read_counts[, c(1, 8:13)], by=\"gene_ID\") %>%\n",
    "                 inner_join(Functional_annotations, by=\"gene_ID\")\n",
    "#Write the table to file\n",
    "write.table(DEG_Ti_counts_FA, file = \"DEG_Output_tables/Ti_Treatment_Control_DGE.txt\", quote = FALSE, row.names = FALSE, sep = \"\\t\")\n",
    "#Display the first 10 rows of the table\n",
    "head(DEG_Ti_counts_FA, n=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Mhzzu3PxAQUx"
   },
   "source": [
    "Example for filtering and counting the Up/Down genes. You can easily filter using the \"filter()\" function just by specifying the dataframe and a logical argument."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Y0VGwBQbAOWD",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "DEG_Ti_up <- filter(DEG_Ti_counts_FA, logFC > 0)\n",
    "print(\"Nr of upregulated genes:\")\n",
    "print(nrow(DEG_Ti_up))\n",
    "DEG_Ti_down <- filter(DEG_Ti_counts_FA, logFC < 0)\n",
    "print(\"Nr of downregulated genes:\")\n",
    "print(nrow(DEG_Ti_down))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can use the \"inner_join\" and the \"anti_join\" functions from the dplyr package to select the common and unique genes for each genotype:\n",
    "\n",
    "`example_file_joined <- inner_join(file1, file2, by=\"gene_ID\")`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ":::{.callout-note title=\"Wrapping up\"}\n",
    "\n",
    "In this notebook, you have worked with RNA-seq results for two white clover genotypes exposed to one night of cold treatment, aiming to identify genes that change their expression in response to the cold treatment. You have learned to perform exploratory data analysis of raw and normalized RNA-seq read counts. You have also performed differential gene expression using `edgeR`. \n",
    "\n",
    "Do you want to look for more analysis? We have an introductory course on bulkRNA analysis at the Danish Health Data Science Sandbox (held periodically in Copenhagen, [keep an eye on the webpage where they list courses](https://heads.ku.dk/course/)). You can also find the material on the  **Transcriptomics Sandbox** on uCloud (for danish users of uCloud), otherwise the course is documented in web format at the respective webpage.\n",
    "\n",
    "&nbsp;\n",
    "\n",
    " <p align=\"center\">\n",
    "  <a href=\"https://cloud.sdu.dk/app/jobs/create?app=transcriptomics\" style=\"background-color: #4266A1; color: #FFFFFF; padding: 20px 20px; text-decoration: none; border-radius: 5px;\">\n",
    "    Transcriptomics Sandbox\n",
    "  </a>\n",
    "</p>\n",
    "  <br>\n",
    "  <br>\n",
    "\n",
    "<p align=\"center\">\n",
    "  <a href=\"https://hds-sandbox.github.io/bulk_RNAseq_course/\" style=\"background-color: #4266A1; color: #FFFFFF; padding: 20px 20px; text-decoration: none; border-radius: 5px;\">\n",
    "    bulkRNA analysis course\n",
    "  </a>\n",
    "</p>\n",
    "  <br>\n",
    "  <br>\n",
    "\n",
    ":::"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "RNAseq_final_May2022.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "NGS (R)",
   "language": "R",
   "name": "ngs_r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  },
  "toc-autonumbering": true,
  "toc-showmarkdowntxt": false,
  "toc-showtags": false,
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
